File: Linked List/142.linked-list-cycle-ii.py
#
# @lc app=leetcode id=142 lang=python3
#
# [142] Linked List Cycle II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return None
        first, second = head, head
        while second:
            first = first.next
            second = second.next
            if second:
                second = second.next
            else:
                return None
            if second == first:
                first = head
                while first != second:
                    first = first.next
                    second = second.next
                return first
        return None    
# @lc code=end

File: Linked List/142.linked-list-cycle-ii.cpp
/*
 * @lc app=leetcode id=142 lang=cpp
 *
 * [142] Linked List Cycle II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(!head || !head->next) return 0;
        auto first = head, second = head;
        while(second)
        {
            first = first->next;
            second = second->next;
            if(second) second = second->next;
            else return 0;
            if(first == second)
            {
                first = head;
                while(first != second)
                {
                    first = first->next;
                    second = second->next;
                }
                return first;
            }
        }
        return 0;
    }
};
// @lc code=end

